### PR TITLE
Added Mercurial SCM support to mix

### PR DESCRIPTION
**Pull request by amclain**
Mercurial (Hg) dependencies can be specified like this:
```elixir
defp deps do
[
  {:foo, hg: "ssh://hg@bitbucket.org/amclain/foo"},
  {:bar, bitbucket: "amclain/bar"}
]
end
```

### CODE REVIEW

File: lib/mix/lib/mix/scm/hg.ex [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  defp get_rev_info do
+    destructure [origin, rev],
+      :os.cmd('hg paths default && hg head -T "{node}\n"')
+      |> IO.iodata_to_binary
+      |> String.split("\n", trim: true)
+    [origin: origin, rev: rev]
+  end
+  
+  defp update_origin(location) do
+    hgrc_path = ".hg/hgrc"
```
Comment (josevalim) [inline lib/mix/lib/mix/scm/hg.ex]: Is this the best way to update the origin? It seems a bit brittle. Doesn't it have command line support?

File: lib/mix/lib/mix/scm/hg.ex [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    {:hg, opts[:hg], rev_info[:rev], get_lock_opts(opts)}
+  end
+  
+  defp get_lock_rev({:hg, _repo, lock, _opts}) when is_binary(lock), do: lock
+  defp get_lock_rev(_), do: nil
+  
+  defp get_lock_opts(opts) do
+    lock_opts = Enum.find_value [:branch, :ref, :tag], &List.keyfind(opts, &1, 0)
+    lock_opts = List.wrap(lock_opts)
+    if opts[:submodules] do
```
Comment (josevalim) [inline lib/mix/lib/mix/scm/hg.ex]: Does mercurial support submodules? I thought they were called subrepos or something?

File: lib/mix/lib/mix/scm/hg.ex [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      _ ->
+        nil
+    end
+  end
+  
+  def accepts_options(_app, opts) do
+    cond do
+      repo = opts[:bitbucket] ->
+        opts |> Keyword.delete(:bitbucket) |> Keyword.put(:hg, "https://bitbucket.org/#{repo}")
+      repo = opts[:hg_github] ->
```
Comment (josevalim) [inline lib/mix/lib/mix/scm/hg.ex]: We don't need to support `hg_github` as it is very uncommon. :)

**Comment by josevalim** [comment]
Thank you. I have added some comments. We also need at least one integration test, please take a look at deps.git_test file.

File: lib/mix/test/mix/scm/hg_test.exs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+Code.require_file "../../test_helper.exs", __DIR__
```
Comment (lexmag) [inline lib/mix/test/mix/scm/hg_test.exs]: I assume that require for test_helper isn't needed here anymore.

File: lib/mix/lib/mix/scm/hg.ex (hunk 1 above)
Comment (amclain) [inline lib/mix/lib/mix/scm/hg.ex]: Unfortunately I wasn't able to find a way to do this using a command built into Hg. Citing [git hg rosetta stone](https://github.com/sympy/sympy/wiki/Git-hg-rosetta-stone):

```text
Git:
git remote add name url # see "git help remote" for more info how to edit paths;
alternatively, you can edit them by hand in .git/config too

Hg:
editing paths in .hg/hgrc
```

I'm all for updating this code later if the feature is added to Hg.

File: lib/mix/lib/mix/scm/hg.ex (hunk 2 above)
Comment (amclain) [inline lib/mix/lib/mix/scm/hg.ex]: You're right, they're called subrepos. I can change the terminology.

File: lib/mix/lib/mix/scm/hg.ex (hunk 3 above)
Comment (amclain) [inline lib/mix/lib/mix/scm/hg.ex]: No problem, I can remove this. My team and I use hg-git to work on Github repos, so I tried to pick an obscure but intuitive keyword that supports this.

**Comment by amclain** [comment]
Whoops, I didn't notice deps.git_test.exs. I can certainly add integration tests.

**Comment by josevalim** [comment]
Regarding the integration tests, you don't need to copy the whole suite. Those tests should be enough:

* https://github.com/elixir-lang/elixir/blob/master/lib/mix/test/mix/tasks/deps.git_test.exs#L44-L84
* https://github.com/elixir-lang/elixir/blob/master/lib/mix/test/mix/tasks/deps.git_test.exs#L181-L275

Apparently we don't test the submodules feature anywhere. :S

**Comment by amclain** [comment]
I’ve added the integration tests and documented the options in deps.ex. It’s been a few years since I’ve used Hg subrepos so I tested them out again. I decided to remove the `:subrepos` keyword because Hg’s default behavior is to pull and update subrepos to the changesets specified by the parent repo.

Also, how does backporting this patch to v1.0 work? Is that something you take care of, or should I submit a PR to the v1.0 branch when this is accepted?

Thanks for your help!

**Comment by josevalim** [comment]
Thank you @amclain!

We are not backporting this to 1.0 because 1.0 is bug fixes only. However, if you need to use this today, you could do something like this on top of your `mix.exs` file:

```elixir
unless Code.ensure_loaded?(Mix.SCM.Mercurial) do
  Code.require_file "scm/mercurial.exs", __DIR__
end
```

Then in `scm/mercurial.exs` you have the Mix.SCM.Mercurial module of this pull request with this extra line at the end:

```elixir
Mix.SCM.prepend Mix.SCM.Mercurial
```

And you should be good to go!

This would actually be a great help because you would be able to give us feedback if everything works as you expect. Another alternative is to make it an archive. You can install it in your path and then on your mix.exs file:

```elixir
if Code.ensure_loaded?(Mix.SCM.Mercurial) do
  Mix.SCM.prepend Mix.SCM.Mercurial
else
  Mix.raise "Mix.SCM.Mercurial not available. Please install it with `mix archive path/to/archive.ez`"
end
```

**Comment by amclain** [comment]
Thanks for the tips! We've implemented the `Code.require_file` suggestion and so far it's working as expected.

**Comment by jcspencer** [comment]
@amclain 
How is this code going? Have you found any issues with it thus far?

I'd really love to see Mercurial support in mix :+1:

**Comment by amclain** [comment]
The code is complete and seems to be working fine so far. If you're able to copy `scm/mercurial.exs` and try it out via @josevalim's instructions, I'd be interested to hear what you think.

**Comment by amclain** [comment]
@josevalim this PR is ready to be merged. Is there anything else you would like me to do?

**Comment by josevalim** [comment]
It looks great! I will merge it whenever I get to focus my efforts on Elixir again. :)

**Comment by josevalim** [comment]
Folks,

After much thinking, I have decided to not merge this feature. The mix test suite is already growing in complexity and this will quite a bit to it. However I do want to this easily accessible. So I recommend you folks to publish an archive. Just move this code to another project, as well as the tests, and run `mix archive.build`. After that, you can make the archive available anywhere, it could even be on Github release pages and do in your mix.exs files:

```elixir
if Code.ensure_loaded?(Mix.SCM.Mercurial) do
  Mix.SCM.prepend Mix.SCM.Mercurial
else
  Mix.raise "Mercurial SCM is not available. Please install it with `mix archive.install https://github.com/path/to/archive.ez`"
end
```

I really appreciate the contribution and, of course, I am available for any questions.

**Comment by jemc** [comment]
Will the Git SCM module also be moved to an archive (for consistency)?

**Comment by josevalim** [comment]
@jemc no it won't because we need at least one remote source in Mix itself to test the whole system (notice Hex *is* an archive though).

**Pull request closed without merging**